File: Python/Rectangle Area.py
class Solution:
    # @param {integer} A
    # @param {integer} B
    # @param {integer} C
    # @param {integer} D
    # @param {integer} E
    # @param {integer} F
    # @param {integer} G
    # @param {integer} H
    # @return {integer}
    def computeArea(self, A, B, C, D, E, F, G, H):
        res = (D-B) * (C-A) + (G-E) * (H-F)
        lb = max(A,E)
        rb = min(C,G)
        ub = min(D,H)
        db = max(B,F)
        return res - max(0, rb-lb) * max(0, ub-db)

File: java/Rectangle_Area.java
// Math
// We can calculate the xGap and yGap between the two rectangle:
// yGap = Math.min(H,D) < Math.max(B,F) ? 0 : Math.min(H,D) - Math.max(B,F);
// xGap = Math.min(G,C) < Math.max(E,A) ? 0 : Math.min(G,C) - Math.max(E,A);
// So, the cross area is xGap*yGap.
// Time and space is both O(1).

public class Solution {
    public int computeArea(int A, int B, int C, int D, int E, int F, int G, int H) {
        return (C-A)*(D-B) + (G-E)*(H-F) - 
               (Math.min(H,D) < Math.max(B,F) ? 0 : Math.min(H,D) - Math.max(B,F))*
               (Math.min(G,C) < Math.max(E,A) ? 0 : Math.min(G,C) - Math.max(E,A));
    }
}
